Spec/Auth.swift

```diff
@@ -3619,5 +3619,331 @@ class Auth : QuickSpec {
+                        options.authParams?.append(URLQueryItem(name: "keySecret", value: keys["keySecret"]) as URLQueryItem)
+                        options.authParams?.append(URLQueryItem(name: "expiresIn", value: String(UInt(tokenDuration))) as URLQueryItem)
+                        let client = ARTRealtime(options: options)
+                        client.setTransport(TestProxyTransport.self)
+                        defer { client.dispose(); client.close() }
+                        
+                        waitUntil(timeout: testTimeout) { done in
+                            client.connection.once(.connected) { stateChange in
+                                let originalToken = client.auth.tokenDetails?.token
+                                let transport = client.transport as! TestProxyTransport
+                                
+                                client.connection.once(.update) { stateChange in
+                                    expect(transport.protocolMessagesReceived.filter({ $0.action == .auth })).to(haveCount(1))
+                                    expect(originalToken).toNot(equal(client.auth.tokenDetails?.token))
+                                    done()
+                                }
+                            }
+                            client.connect()
+                        }
+                    }
+                }
+            }
+
+            // RSA8g
+            context("when using authCallback") {
+                let options = AblyTests.clientOptions()
+
+                context("with valid credentials") {
+                    options.authCallback = { tokenParams, completion in
+                        let token = ARTTokenDetails(token: getJWTToken()!)
+                        completion(token, nil)
+                    }
+                    let client = ARTRealtime(options: options)
+                    defer { client.dispose(); client.close() }
+
+                    it("pulls stats successfully") {
+                        waitUntil(timeout: testTimeout) { done in
+                            client.stats{ stats, error in
+                                expect(error).to(beNil())
+                                done()
+                            }
+                        }
+                    }
+                }
+
+                context("with invalid credentials") {
+                    options.authCallback = { tokenParams, completion in
+                        let token = ARTTokenDetails(token: getJWTToken(invalid: true)!)
+                        completion(token, nil)
+                    }
+                    let client = ARTRealtime(options: options)
+                    defer { client.dispose(); client.close() }
+
+                    it("fails to connect") {
+                        waitUntil(timeout: testTimeout) { done in
+                            client.connection.once(.failed) { stateChange in
```

Comment:
I thought we had agreed that https://docs.ably.io/client-lib-development-guide/features/#RSA4b stated that the connection should not become failed, but instead should become disconnected and retry. I believe @paddybyers updated the spec after chatting with you about this at some point, although oddly I can't find a matching commit.

Reply:
Add this as a TODO above

Reply:
@mattheworiordan RSA4b says that a REST request fails if a token error is returned after a retry, which is correct, and is unrelated to this test.

https://docs.ably.io/client-lib-development-guide/features/#RTN14b relates to token failures on connection creation - it says that a connection should not fail in that case, but become disconnected. This test is therefore wrong and, because it passes, there is also a bug in the library (which will have existed before these JWT changes). @funkyboy pls raise a separate issue for that (and find the other existing, non-JWT, invalid test).

The discussion you remembered related was https://github.com/ably/docs/issues/429, relating to RTN15h, and that's to do with failures to reconnect (ie not initial connection failures). I'm guessing that there's going to be an issue for iOS there as well. I will do a docs PR for that.

Reply:
@paddybyers @mattheworiordan The non-JWT corresponding test is https://github.com/ably/ably-ios/blob/develop/Spec/Auth.swift#L305 and again it tests for `FAILED`. 

Issue raised here: https://github.com/ably/ably-ios/issues/730

Reply:
@paddybyers [This](https://github.com/ably/ably-ios/blob/develop/Spec/Auth.swift#L216) seems to test the new `RTN15h1` that you wrote. 
[These](https://github.com/ably/ably-ios/blob/develop/Spec/Auth.swift#L190-L332) all riff around the same "token expired/wrong" conditions but none seems to tests for the "single attempt to renew the token" clause.

Should I open a new issue about this?

cc @mattheworiordan

Reply:
> @mattheworiordan RSA4b says that a REST request fails if a token error is returned after a retry, which is correct, and is unrelated to this test.

Oops, I should have scrolled up to see the context of the test I found.  That explains why I could not find the record of the change we made in the spec!

Reply:
@paddybyers 
> ably/docs#439. @funkyboy can you check that behaviour please in this library and see if complies with the DISCONNECTED or FAILED behaviour?

Opened issue here #731